File: stations/hook-observatory/handlers/secret_scan.py
"""
Secret scanning before git push — PreToolUse handler for Bash tool.

CRITICAL handler: always runs (not subject to 5s deferrable budget).
Scans git diff for hardcoded secrets before push. Only fires on `git push`.

Escape hatches:
  - SECRET_SCAN_DISABLE=1 env var disables entirely
  - `# nosec` at end of line skips that line
  - 3s timeout guard: warns instead of blocking on timeout
"""

from __future__ import annotations

import os
import re

from .base import ALLOW, HookResult, block, message, run_cmd

# ---------------------------------------------------------------------------
# Secret patterns — tuned for low false-positive rate
# ---------------------------------------------------------------------------

_PATTERNS: list[tuple[re.Pattern, str]] = [
    (re.compile(r"AKIA[0-9A-Z]{16}"), "AWS Access Key"),
    (re.compile(r"gh[ps]_[A-Za-z0-9_]{36,}"), "GitHub Token"),
    (re.compile(r"xox[baprs]-[A-Za-z0-9-]{10,}"), "Slack Token"),
    (re.compile(r"-----BEGIN .* PRIVATE KEY-----"), "Private Key"),  # nosec
    (
        re.compile(
            r"(?:api[_-]?key|secret[_-]?key|access[_-]?token|password)"
            r"\s*[=:]\s*['\"][^'\"]{12,}",
            re.IGNORECASE,
        ),
        "Generic Secret Assignment",
    ),
]

# Lines containing these tokens are likely examples/placeholders, skip them
_FALSE_POSITIVE_TOKENS = {
    "example",
    "placeholder",
    "changeme",
    "todo",
    "fixme",
    "your-",
    "your_",
    "xxx",
    "dummy",
    "fake",
    "mock",
    "test_",
    "sample",
    "template",
}

# File paths to skip (test fixtures, examples, docs)
_SKIP_PATH_PATTERNS = {
    "test/",
    "tests/",
    "fixtures/",
    "mocks/",
    ".example",
    ".sample",
    ".template",
}


def handle(event_type: str, tool_name: str, tool_input: dict, raw_input: str) -> HookResult:
    if tool_name != "Bash":
        return ALLOW

    # Escape hatch: env var disable
    if os.environ.get("SECRET_SCAN_DISABLE") == "1":
        return ALLOW

    command = tool_input.get("command", "")
    if not command:
        return ALLOW

    # Only fire on git push (not --dry-run)
    if not re.search(r"\bgit\s+push\b", command):
        return ALLOW
    if "--dry-run" in command:
        return ALLOW

    # Get diff of unpushed commits
    diff_text = _get_unpushed_diff()
    if diff_text is None:
        return message(
            "\u26a0\ufe0f secret scan: \u7121\u6cd5\u53d6\u5f97 diff\uff0c\u6383\u63cf\u8df3\u904e"
        )

    if not diff_text.strip():
        return ALLOW

    # Scan only added lines
    findings = _scan_added_lines(diff_text)
    if findings:
        detail = "\n".join(f"  - {f}" for f in findings[:5])
        if len(findings) > 5:
            detail += f"\n  ... and {len(findings) - 5} more"
        return block(
            f"\u5075\u6e2c\u5230\u7591\u4f3c secret ({len(findings)} \u8655):\n{detail}\n"
            f"\u82e5\u70ba\u8aa4\u5831\uff0c\u5728\u884c\u5c3e\u52a0 # nosec"
            f" \u6216\u8a2d SECRET_SCAN_DISABLE=1"
        )

    return ALLOW


def _get_unpushed_diff() -> str | None:
    """Get diff of commits not yet pushed. Returns None on failure/timeout."""
    # Try upstream tracking branch first
    result = run_cmd(
        ["git", "diff", "@{upstream}..HEAD"],
        timeout=3,
    )
    if result and result.returncode == 0:
        return result.stdout

    # Fallback: diff against origin/main
    result = run_cmd(
        ["git", "diff", "origin/main..HEAD"],
        timeout=3,
    )
    if result and result.returncode == 0:
        return result.stdout

    # Last resort: just the latest commit
    result = run_cmd(
        ["git", "log", "-1", "-p", "--format="],
        timeout=3,
    )
    if result and result.returncode == 0:
        return result.stdout

    return None


def _scan_added_lines(diff_text: str) -> list[str]:
    """Scan added lines in unified diff for secret patterns."""
    findings: list[str] = []
    current_file = ""

    for line in diff_text.splitlines():
        # Track current file from diff headers
        if line.startswith("+++ b/"):
            current_file = line[6:]
            continue

        # Skip non-added lines
        if not line.startswith("+") or line.startswith("+++"):
            continue

        added_line = line[1:]  # strip the leading +

        # Skip if file path matches skip patterns
        if any(pat in current_file for pat in _SKIP_PATH_PATTERNS):
            continue

        # Skip comments
        stripped = added_line.strip()
        if stripped.startswith(("#", "//", "/*", "*", "<!--")):
            continue

        # Skip lines with # nosec annotation
        if "# nosec" in added_line or "// nosec" in added_line:
            continue

        # Skip false-positive tokens
        lower = added_line.lower()
        if any(token in lower for token in _FALSE_POSITIVE_TOKENS):
            continue

        # Check against secret patterns
        for pattern, label in _PATTERNS:
            if pattern.search(added_line):
                # Truncate the line for display (don't leak the actual secret)
                preview = added_line.strip()[:40]
                if len(added_line.strip()) > 40:
                    preview += "..."
                findings.append(f"{label} in {current_file}: {preview}")
                break  # one finding per line is enough

    return findings

File: internal/handlers/secret_scan.go
package handlers

import (
	"os"
	"regexp"
	"strings"
	"time"

	"github.com/joneshong/hook-dispatcher/internal/core"
)

func init() {
	core.Register("PreToolUse", core.Entry{
		Matcher:    "Bash",
		Handler:    secretScanHandle,
		Critical:   true,
		ModuleName: "secret_scan",
	})
}

// secretScanHandle is the Go port of handlers/secret_scan.py.
//
// Scans the unpushed git diff for hardcoded secrets before a `git push`.
// Escape hatches: SECRET_SCAN_DISABLE=1 env var, or `# nosec` / `// nosec` line suffix.
// On timeout or diff-fetch failure, warns (message) rather than blocking.
func secretScanHandle(_, toolName string, toolInput map[string]any, _ string) core.HookResult {
	if toolName != "Bash" {
		return core.Allow()
	}
	if os.Getenv("SECRET_SCAN_DISABLE") == "1" {
		return core.Allow()
	}

	command, _ := toolInput["command"].(string)
	if command == "" {
		return core.Allow()
	}

	// Only fire on `git push` (not --dry-run)
	if !reGitPushCmd.MatchString(command) {
		return core.Allow()
	}
	if strings.Contains(command, "--dry-run") {
		return core.Allow()
	}

	diffText, ok := ssGetUnpushedDiff()
	if !ok {
		return core.Message("⚠️ secret scan: 無法取得 diff，掃描跳過")
	}
	if strings.TrimSpace(diffText) == "" {
		return core.Allow()
	}

	findings := ssScanAddedLines(diffText)
	if len(findings) == 0 {
		return core.Allow()
	}

	var sb strings.Builder
	limit := len(findings)
	if limit > 5 {
		limit = 5
	}
	for _, f := range findings[:limit] {
		sb.WriteString("  - ")
		sb.WriteString(f)
		sb.WriteByte('\n')
	}
	if len(findings) > 5 {
		sb.WriteString("  ... and ")
		sb.WriteString(itoa(len(findings) - 5))
		sb.WriteString(" more\n")
	}

	return core.Block(
		"偵測到疑似 secret (" + itoa(len(findings)) + " 處):\n" +
			sb.String() +
			"若為誤報，在行尾加 # nosec 或設 SECRET_SCAN_DISABLE=1",
	)
}

// ---------------------------------------------------------------------------
// Secret patterns — mirrors Python _PATTERNS exactly
// ---------------------------------------------------------------------------

type ssPattern struct {
	re    *regexp.Regexp
	label string
}

var ssPatterns = []ssPattern{
	{regexp.MustCompile(`AKIA[0-9A-Z]{16}`), "AWS Access Key"},
	{regexp.MustCompile(`gh[ps]_[A-Za-z0-9_]{36,}`), "GitHub Token"},
	{regexp.MustCompile(`xox[baprs]-[A-Za-z0-9-]{10,}`), "Slack Token"},
	{regexp.MustCompile(`-----BEGIN .* PRIVATE KEY-----`), "Private Key"}, // nosec
	// Generic secret assignment — case-insensitive
	{regexp.MustCompile(`(?i)(?:api[_-]?key|secret[_-]?key|access[_-]?token|password)\s*[=:]\s*['"][^'"]{12,}`), "Generic Secret Assignment"},
}

// False-positive tokens — lines containing any of these are skipped
var ssFalsePositiveTokens = []string{
	"example", "placeholder", "changeme", "todo", "fixme",
	"your-", "your_", "xxx", "dummy", "fake", "mock", "test_", "sample", "template",
}

// File path substrings that mean "skip this file"
var ssSkipPathPatterns = []string{
	"test/", "tests/", "fixtures/", "mocks/",
	"_archive/", "/archive/",
	".example", ".sample", ".template",
}

// Comment prefixes that mean "skip this line"
var ssCommentPrefixes = []string{"#", "//", "/*", "*", "<!--"}

var reGitPushCmd = regexp.MustCompile(`\bgit\s+push\b`)

// ---------------------------------------------------------------------------
// Diff fetching
// ---------------------------------------------------------------------------

// ssGetUnpushedDiff returns (diffText, true) or ("", false) on failure.
func ssGetUnpushedDiff() (string, bool) {
	attempts := [][]string{
		{"git", "diff", "@{upstream}..HEAD"},
		{"git", "diff", "origin/main..HEAD"},
		{"git", "log", "-1", "-p", "--format="},
	}
	for _, args := range attempts {
		res := core.RunCmd(args, "", 3*time.Second, "")
		if res != nil && res.ExitCode == 0 {
			return res.Stdout, true
		}
	}
	return "", false
}

// ---------------------------------------------------------------------------
// Line scanning
// ---------------------------------------------------------------------------

func ssScanAddedLines(diffText string) []string {
	var findings []string
	currentFile := ""

	for _, rawLine := range strings.Split(diffText, "\n") {
		// Track current file
		if strings.HasPrefix(rawLine, "+++ b/") {
			currentFile = rawLine[6:]
			continue
		}
		// Only added lines (not the +++ header itself)
		if !strings.HasPrefix(rawLine, "+") || strings.HasPrefix(rawLine, "+++") {
			continue
		}
		addedLine := rawLine[1:] // strip leading +

		// Skip files in skip-path patterns
		if ssMatchesSkipPath(currentFile) {
			continue
		}

		stripped := strings.TrimSpace(addedLine)

		// Skip comment lines
		if ssIsComment(stripped) {
			continue
		}

		// Skip nosec annotations
		if strings.Contains(addedLine, "# nosec") || strings.Contains(addedLine, "// nosec") {
			continue
		}

		// Skip false-positive tokens
		lower := strings.ToLower(addedLine)
		if ssHasFalsePositive(lower) {
			continue
		}

		// Check patterns
		for _, p := range ssPatterns {
			if p.re.MatchString(addedLine) {
				preview := strings.TrimSpace(addedLine)
				if len(preview) > 40 {
					preview = preview[:40] + "..."
				}
				findings = append(findings, p.label+" in "+currentFile+": "+preview)
				break // one finding per line
			}
		}
	}
	return findings
}

func ssMatchesSkipPath(file string) bool {
	for _, pat := range ssSkipPathPatterns {
		if strings.Contains(file, pat) {
			return true
		}
	}
	return false
}

func ssIsComment(stripped string) bool {
	for _, pfx := range ssCommentPrefixes {
		if strings.HasPrefix(stripped, pfx) {
			return true
		}
	}
	return false
}

func ssHasFalsePositive(lower string) bool {
	for _, tok := range ssFalsePositiveTokens {
		if strings.Contains(lower, tok) {
			return true
		}
	}
	return false
}

// ---------------------------------------------------------------------------
// Tiny helpers
// ---------------------------------------------------------------------------

func itoa(n int) string {
	if n == 0 {
		return "0"
	}
	buf := [20]byte{}
	pos := len(buf)
	neg := n < 0
	if neg {
		n = -n
	}
	for n > 0 {
		pos--
		buf[pos] = byte('0' + n%10)
		n /= 10
	}
	if neg {
		pos--
		buf[pos] = '-'
	}
	return string(buf[pos:])
}
